sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,310 @@
-     *
-     * Example:
-     *
-     *  list =                      [string, string, long]
-     *  varArgsStartingPosition =   1
-     *
-     *     -> variable sequence =   [string, long]
-     *
-     * Matching lists:
-     *
-     *  [string, string, long]
-     *  [string, string, long, string, long]
-     *  ...
-     *
-     * If the value is -1, no variable argument matching is done.
+     * Creates an operator which takes its matchers from an iterator. The size of the
+     * validated arguments is not checked, however it needs to be greater or equal to the
+     * number of matchers returned by the iterator;
      */
-    private final int varArgsStartingPosition;
+    public static SignatureOperator of(Iterable<ArgMatcher> argMatchers) {
+        return new IterableArgs(argMatchers);
+    }
+
+    public static SignatureOperator of(ArgMatcher... matchers) {
+        if (matchers.length == 0) {
+            return EMPTY;
+        }
+        return new VarArgs(matchers);
+    }
+
+    public static SignatureOperator size(int minSize, int maxSize) {
+        Preconditions.checkArgument(minSize <= maxSize, "minSize needs to be smaller than maxSize");
+        return dataTypes -> (dataTypes.size() >= minSize && dataTypes.size() <= maxSize) ? dataTypes : null;
+    }
+
 
-    public Signature(Collection<? extends DataType> c) {
-        this(-1, c);
+    public static SignatureOperator of(boolean hasVarArgs, boolean strictVarArgTypes, ArgMatcher... matchers) {
+        return new VarArgs(hasVarArgs, strictVarArgTypes, matchers);
     }
 
-    public Signature(int varArgsStartingPosition, Collection<? extends DataType> c) {
-        assert varArgsStartingPosition < c.size() : "varArgs starting position exceeds args definition";
-        this.varArgsStartingPosition = varArgsStartingPosition;
-        list = ImmutableList.copyOf(c);
+
+    public static SignatureOperator of(int numArgs) {
+        Preconditions.checkArgument(numArgs >= 0, "numArgs must not be negative");
+        if (numArgs == 0) {
+            return EMPTY;
+        }
+        return in -> in.size() == numArgs ? in : null;
     }
 
-    public Signature(DataType... elements) {
-        this(-1, elements);
+    public static SignatureOperator of(DataType... dataTypes) {
+        if (dataTypes.length == 0) {
+            return EMPTY;
+        }
+        ArgMatcher[] matchers = new ArgMatcher[dataTypes.length];
+        for (int i = 0; i < matchers.length; i++) {
+            matchers[i] = ArgMatcher.rewriteTo(dataTypes[i]);
+        }
+        return new VarArgs(matchers);
     }
 
-    public Signature(int varArgsStartingPosition, DataType... elements) {
-        assert varArgsStartingPosition < elements.length : "varArgs starting position exceeds args definition";
-        this.varArgsStartingPosition = varArgsStartingPosition;
-        list = ImmutableList.copyOf(elements);
+    public static SignatureOperator of(List<DataType> dataTypes) {
```

Comment:
Could one of these overloads (array/list) be removed, or just call into the other? Since this stuff is mostly called on node startup I don't think we have to micro optimize to avoid a single "asLists" allocation here.